### PR TITLE
BaSP-TG-41: Admin

### PR DESCRIPTION
**Pull request by GonzaloAsencio**
**List Admin :**

- Retrieve the data from the data base and list all the admins.
- Create a list view where each item has a button to add, update and delete.

**Delete an admin:**

-  Delete an admin from the list.

**Create an admin:**

- Create the form view.
- Create a new admin.

**Update an admin:**

- Update an admin from the list.

### CODE REVIEW

**Comment by GonzaloAsencio** [comment]
I use "setTimeout" for the warnings, because I don't finish on time.

**Comment by Ignacio-Sormani** [comment]
Please resolve conflicts

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,2 @@
+{
```
Comment (Ignacio-Sormani) [inline .vscode/settings.json:1]: I think you should remove this

File: package.json [hunk 2]
```diff
@@ -6,6 +6,7 @@
     "@testing-library/jest-dom": "^5.15.1",
     "@testing-library/react": "^11.2.7",
     "@testing-library/user-event": "^12.8.3",
+    "cors": "^2.8.5",
```
Comment (Ignacio-Sormani) [inline package.json:9]: We don't need cors for react app

File: src/Components/Admins/Modal/modal.module.css [hunk 3]
```diff
@@ -0,0 +1,43 @@
+.moduleContainer {
+    position: fixed;
+    left: 0;
+    top: 0;
+    right: 0;
+    bottom: 0;
+    justify-content: center;
+    align-items: center;
+    background-color: rgba(0,0,0,0.5);
+  }
```
Comment (Ignacio-Sormani) [inline src/Components/Admins/Modal/modal.module.css:10]: Please leave a blank line between each css block

File: src/Components/Admins/Modal/modalAdmin.jsx [hunk 4]
```diff
@@ -0,0 +1,33 @@
+import styles from './modal.module.css';
+
+function Modal(props) {
+  if (!props.show) {
+    return null;
+  }
+  const onCloseModal = () => {
```
[the remark this answers is not in the document]
Comment (Ignacio-Sormani) [reply src/Components/Admins/Modal/modalAdmin.jsx:7]: Also leave one blank line between functions

File: src/Components/Admins/adminRow.jsx [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+  const openModal = () => {
+    setModal(true);
+  };
+
+  const closeModal = () => {
+    setModal(false);
+  };
+
+  const onCloseModal = () => {
```
Comment (Ignacio-Sormani) [inline src/Components/Admins/adminRow.jsx:16]: Instead of onCloseModal it could be called deleteAdmin

Comment (Ignacio-Sormani) [reply src/Components/Admins/adminRow.jsx:16]: Use async await and remove useTimeout

File: src/Components/Admins/adminRow.jsx [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    <tr>
+      <td>{listAdmin._id}</td>
+      <td>{listAdmin.name}</td>
+      <td>{listAdmin.lastName}</td>
+      <td>{listAdmin.email}</td>
+      <td>{listAdmin.password}</td>
+      <td>
+        <a href={`admins/form?id=${listAdmin._id}`}>
+          <button>
+            <img src={`${process.env.PUBLIC_URL}/assets/images/iconEdit.svg`} />
```
Comment (Ignacio-Sormani) [inline src/Components/Admins/adminRow.jsx:39]: We don't hace an env variable called PUBLIC_URL. Just use the string for now

Also, if you want to set a new env variable in react it should have a prefix: "REACT_APP_XXXX_XXXX"

File: src/Components/Admins/adminRow.jsx [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        <Modal
+          show={showModal}
+          closeModal={closeModal}
+          onCloseModal={onCloseModal}
+          id={listAdmin._id}
+          title={'DELETE ADMIN'}
+          text={`Do you want to remove ${listAdmin.name}?`}
+        />
+        <button onClick={openModal}>
+          <img src={`${process.env.PUBLIC_URL}/assets/images/iconTrash.svg`} />
```
Comment (Ignacio-Sormani) [inline src/Components/Admins/adminRow.jsx:53]: Same comment as above

File: src/Components/Admins/admins.module.css [hunk 8]
```diff
@@ -1,6 +1,116 @@
...
   justify-content: center;
   align-items: center;
-}
\ No newline at end of file
+  width: 500px;
+  background-color: white;
+  border-radius: 5px;
+  border: solid 1px grey;
+  margin: auto;
+
```
Comment (Ignacio-Sormani) [inline src/Components/Admins/admins.module.css:74]: Remove this empty line

File: src/Components/Admins/admins.module.css [hunk 9]
```diff
@@ -1,6 +1,116 @@
...
+  font-weight: 800;
+  align-items: center;
+  color: white;
+  background-color: #373867 ;
+  border-style: none;
+  border: 5px;
+  border-radius: 5px;
+}
+
+
```
Comment (Ignacio-Sormani) [inline src/Components/Admins/admins.module.css:104]: Just leave one empty line here

File: src/Components/Admins/editAdminForm.jsx [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+function Form() {
+  const [inputValue, setInputValue] = useState({
+    name: '',
+    lastName: '',
+    email: '',
+    password: ''
+  });
+
+  useEffect(async () => {
+    const params = new URLSearchParams(window.location.search);
```
Comment (Ignacio-Sormani) [inline src/Components/Admins/editAdminForm.jsx:13]: Change this to read the path param from window.location.pathname. You can look at the code merged in TG-50, file: "Projects/Form/index.js"

File: src/Components/Admins/editAdminForm.jsx [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+      }
+    }
+  }, []);
+
+  const onChange = (e) => {
+    setInputValue({ ...inputValue, [e.target.name]: e.target.value });
+  };
+
+  const onSubmit = () => {
+    const params = new URLSearchParams(window.location.search);
```
Comment (Ignacio-Sormani) [inline src/Components/Admins/editAdminForm.jsx:39]: Same as above

File: src/Components/Admins/index.js [hunk 12]
```diff
@@ -1,10 +1,32 @@
...
+
   return (
-    <section className={styles.container}>
-      <h2>Admins</h2>
-    </section>
+    <div className={styles.container}>
+      <div className={styles.adminHeader}>
+        <h2>Admin</h2>
+        <a href={`admins/form`}>
+          <button>+</button>
```
Comment (Ignacio-Sormani) [inline src/Components/Admins/index.js:25]: Should not use a block element inside an inline element. Just don't use the button

**Review by Ignacio-Sormani — CHANGES_REQUESTED** [review]

File: .vscode/settings.json (hunk 1 above)
Comment (GonzaloAsencio) [reply .vscode/settings.json:1]: Fixed!

File: package.json (hunk 2 above)
Comment (GonzaloAsencio) [reply package.json:9]: Removed!

File: src/Components/Admins/Modal/modal.module.css (hunk 3 above)
Comment (GonzaloAsencio) [reply src/Components/Admins/Modal/modal.module.css:10]: Done

File: src/Components/Admins/Modal/modalAdmin.jsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (GonzaloAsencio) [reply src/Components/Admins/Modal/modalAdmin.jsx:7]: Done

File: src/Components/Admins/adminRow.jsx (hunk 6 above)
Comment (GonzaloAsencio) [reply src/Components/Admins/adminRow.jsx:39]: ohh i see. Fixed!

File: src/Components/Admins/adminRow.jsx (hunk 7 above)
Comment (GonzaloAsencio) [reply src/Components/Admins/adminRow.jsx:53]: Fixed

File: src/Components/Admins/admins.module.css (hunk 8 above)
Comment (GonzaloAsencio) [reply src/Components/Admins/admins.module.css:74]: Done

File: src/Components/Admins/admins.module.css (hunk 9 above)
Comment (GonzaloAsencio) [reply src/Components/Admins/admins.module.css:104]: Done

**Review by fede230298 — COMMENTED** [review]
Is there a specific reason you are using Promises insted of Async Await when fetching?

**Comment by GonzaloAsencio** [comment]
For nothing in particular.I did it at the beginning of the project and it stayed that way.

File: src/Components/Admins/adminTable.jsx [hunk 13]
```diff
@@ -0,0 +1,34 @@
+import React from 'react';
+import RowAdmin from './adminRow';
+import styles from './admins.module.css';
+
+const AdminTable = ({ list, deleteItem }) => {
+  return (
+    <div className={styles.list}>
+      {list === undefined || list.length === 0 ? (
+        <span>There are no admin yet!</span>
```
Comment (fede230298) [inline src/Components/Admins/adminTable.jsx:9]: Maybe you can invert the elements and put first the table and as only condition 
`{list.length ? (<table>...</table> : <span>There are no admin yet!</span>`

Comment (GonzaloAsencio) [reply src/Components/Admins/adminTable.jsx]: Done!

**Review by fede230298 — APPROVED** [review]
SGTM 🚀

File: src/Components/Layout/index.js [hunk 14]
```diff
@@ -17,10 +18,10 @@ function Layout() {
     case 'admins':
       currentScreen = <Admins />;
       break;
-    case 'admin-form':
-      currentScreen = <Admins />;
+    case 'admins-form':
```
Comment (Ignacio-Sormani) [inline src/Components/Layout/index.js:21]: Path should be admin-form

File: src/Components/Admins/index.js [hunk 15]
```diff
@@ -1,10 +1,30 @@
+import { useEffect, useState } from 'react';
+import Table from './adminTable';
 import styles from './admins.module.css';
 
 function Admins() {
+  const [admins, saveAdmins] = useState([]);
+
+  useEffect(async () => {
+    await fetch(`${process.env.REACT_APP_API_URL}/admins`)
```
Comment (Ignacio-Sormani) [inline src/Components/Admins/index.js:9]: Remove .then and .catch. You should just use async await

File: src/Components/Admins/editAdminForm.jsx [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+        const data = await response.json();
+        setInputValue({
+          name: data.data.name,
+          lastName: data.data.lastName,
+          email: data.data.email,
+          password: data.data.password
+        });
+        return;
+      } catch (err) {
+        setTimeout(() => {
```
Comment (Ignacio-Sormani) [inline src/Components/Admins/editAdminForm.jsx:27]: Remove setTimeout

File: src/Components/Admins/editAdminForm.jsx [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+    if (adminId !== 'admins-form') {
+      const options = {
+        method: 'PUT',
+        headers: {
+          'Content-Type': 'application/json'
+        },
+        body: JSON.stringify(inputValue)
+      };
+      const url = `${process.env.REACT_APP_API_URL}/admins/${adminId}`;
+      fetch(url, options).then(async (response) => {
```
Comment (Ignacio-Sormani) [inline src/Components/Admins/editAdminForm.jsx:50]: Remove .then. You should just use async await

File: src/Components/Admins/editAdminForm.jsx [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+    } else {
+      const options = {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json'
+        },
+        body: JSON.stringify(inputValue)
+      };
+      const url = `${process.env.REACT_APP_API_URL}/admins`;
+      fetch(url, options).then(async (response) => {
```
Comment (Ignacio-Sormani) [inline src/Components/Admins/editAdminForm.jsx:68]: Remove .then. You should just use async await

File: src/Components/Admins/index.js (hunk 15 above)
Comment (Ignacio-Sormani) [inline src/Components/Admins/index.js:9]: Remove .then and .catch. You should just use async await

**Review by Ignacio-Sormani — CHANGES_REQUESTED** [review]

File: src/Components/Layout/index.js (hunk 14 above)
Comment (GonzaloAsencio) [reply src/Components/Layout/index.js:21]: Changed!

File: src/Components/Admins/index.js (hunk 15 above)
Comment (GonzaloAsencio) [reply src/Components/Admins/index.js:9]: Changed!

File: src/Components/Admins/editAdminForm.jsx (hunk 16 above)
Comment (GonzaloAsencio) [reply src/Components/Admins/editAdminForm.jsx:27]: Removed

File: src/Components/Admins/index.js (hunk 15 above)
Comment (GonzaloAsencio) [reply src/Components/Admins/index.js:9]: Removed

File: src/Components/Admins/editAdminForm.jsx (hunk 17 above)
Comment (GonzaloAsencio) [reply src/Components/Admins/editAdminForm.jsx:50]: done

File: src/Components/Admins/editAdminForm.jsx (hunk 18 above)
Comment (GonzaloAsencio) [reply src/Components/Admins/editAdminForm.jsx:68]: done

**Review by Ignacio-Sormani — APPROVED** [review]

**Review by sofihoj — APPROVED** [review]

**Pull request merged**